src/jrd/cch.cpp

```diff
@@ -5312,3 +5247,269 @@ void BufferDesc::unLockIO(thread_db* tdbb)
+
+	chain_type* new_chains = FB_NEW_POOL(m_pool) chain_type[count];
+	m_count = count;
+	m_chains = new_chains;
+
+#ifndef HASH_USE_CDS_LIST
+	// Initialize all new new_chains
+	for (chain_type* que = new_chains; que < new_chains + count; que++)
+		QUE_INIT(*que);
+#endif
+
+	if (!old_chains)
+		return;
+
+	const chain_type* const old_end = old_chains + old_count;
+
+	// Move any active buffers from old hash table to new
+	for (chain_type* old_tail = old_chains; old_tail < old_end; old_tail++)
+	{
+#ifndef HASH_USE_CDS_LIST
+		while (QUE_NOT_EMPTY(*old_tail))
+		{
+			QUE que_inst = old_tail->que_forward;
+			BufferDesc* bdb = BLOCK(que_inst, BufferDesc, bdb_que);
+			QUE_DELETE(*que_inst);
+			QUE mod_que = &new_chains[hash(bdb->bdb_page)];
+			QUE_INSERT(*mod_que, *que_inst);
+		}
+#else
+		while (!old_tail->empty())
+		{
+			auto n = old_tail->begin();
+			old_tail->erase(n->first);				// bdb_page
+
+			chain_type* new_chain = &m_chains[hash(n->first)];
+			new_chain->insert(n->first, n->second);	// bdb_page, bdb
+		}
+#endif
+	}
+
+	delete[] old_chains;
+}
+
+void BCBHashTable::clear()
+{
+	if (!m_chains)
+		return;
+
+#ifdef HASH_USE_CDS_LIST
+	const chain_type* const end = m_chains + m_count;
+	for (chain_type* tail = m_chains; tail < end; tail++)
+		tail->clear();
+#endif
+
+	delete[] m_chains;
+	m_chains = nullptr;
+	m_count = 0;
+}
+
+inline BufferDesc* BCBHashTable::find(const PageNumber& page) const
+{
+	auto& list = m_chains[hash(page)];
+
+#ifndef HASH_USE_CDS_LIST
+	QUE que_inst = list.que_forward;
+	for (; que_inst != &list; que_inst = que_inst->que_forward)
+	{
+		BufferDesc* bdb = BLOCK(que_inst, BufferDesc, bdb_que);
+		if (bdb->bdb_page == page)
+			return bdb;
+	}
+
+#else // HASH_USE_CDS_LIST
+	auto ptr = list.get(page);
+	if (!ptr.empty())
+	{
+		fb_assert(ptr->second != nullptr);
+#ifdef DEV_BUILD
+		// Original libcds have no update(key, value), use this code with it,
+		// see also comment in get_buffer()
+		while (ptr->second == nullptr)
+			cds::backoff::pause();
+#endif
+		if (ptr->second->bdb_page == page)
+			return ptr->second;
+	}
+#endif
+
+	return nullptr;
+}
+
+inline BufferDesc* BCBHashTable::emplace(BufferDesc* bdb, const PageNumber& page, bool remove)
+{
+#ifndef HASH_USE_CDS_LIST
+	// bcb_syncObject should be locked in EX mode
+
+	BufferDesc* bdb2 = find(page);
+	if (!bdb2)
+	{
+		if (remove)
+			QUE_DELETE(bdb->bdb_que);
+
+		que& mod_que = m_chains[hash(page)];
+		QUE_INSERT(mod_que, bdb->bdb_que);
+	}
+	return bdb2;
+#else // HASH_USE_CDS_LIST
+
+	BufferDesc* bdb2 = nullptr;
+	BdbList& list = m_chains[hash(page)];
+
+/*
+	// Original libcds have no update(key, value), use this code with it
+
+	auto ret = list.update(page, [bdb, &bdb2](bool bNew, BdbList::value_type& val)
+		{
+			if (bNew)
+				val.second = bdb;
+			else
+				while (!(bdb2 = val.second))
+					cds::backoff::pause();
+		},
+		true);
+*/
+
+	auto ret = list.update(page, bdb, [&bdb2](bool bNew, BdbList::value_type& val)
```

Comment:
How well maintained is this library?

Your issue from Jul 25, 2021 is not answered?

We depend on custom version, so it's not possible to use system version of library in Linux like we use for others libraries?

Reply:
Library was maintained very well until last year when decision to use was made.
It have good reputation and stability, a lot of tests provided.
I see no problem re (not) usage of system version, we already does it and nobody died so far.